### PR TITLE
chore: fix support for page/category specific `og:image`

### LINKED ISSUE #37: Sharing image is ignored
If you try to share https://webdev.dartlang.org/codelabs/darrrt, it gets the default image (or on-page images), not the one specified in the header:

```
description: "Take your first step to learning Dart fast."
...
snippet_img: images/piratemap.jpg
```

(I think the image in the header is a broken URL, since it needs to be a full path, but still. It should link to https://webdev.dartlang.org/codelabs/darrrt/images/piratemap.jpg.)

### PR DESCRIPTION
**Pull request by chalin**
Staged at https://angulardart-org-dev.firebaseapp.com

- Fixes #37 
- Adds an AngularDart specific og:image consisting of the AngularDart logo. (This might need tweaking --- see below --- but any image tweaking should be done in a followup PR.)
- I've ensured that the Facebook debugger reports no problems; see below for sampled pages.

Note: the `angulardart.png` file was created using imagemagick:
```sh
convert -size 630x630 -extent 1200x630 -gravity center angulardart.svg angulardart.png
```

cc @Sfshaza @filiph 

---
Test output from the Facebook debugger when run over https://angulardart-org-dev.firebaseapp.com/

### Main page

Meta Tag | Value
-- | --
og:url | https://angulardart-org-dev.firebaseapp.com/
og:type | website
og:title | Dart for the web
og:image |  https://angulardart-org-dev.firebaseapp.com/assets/dart-webdev-logo.png
og:description | Dart is a modern language well suited to writing fast, scalable, web apps.

### AngularDart codelab

Meta Tag | Value
-- | --
og:url | https://angulardart-org-dev.firebaseapp.com/codelabs/ng2
og:type | article
og:title | Avast, Ye Pirates: Write an AngularDart App
og:image |  https://angulardart-org-dev.firebaseapp.com/codelabs/ng2/images/piratemap.jpg?3
og:description | Learn how to use Dart + Angular by creating a simple web app.

### ACX codelab

Meta Tag | Value
-- | --
og:url | https://angulardart-org-dev.firebaseapp.com/codelabs/angular_components/3-usebuttons
og:type | article
og:title | Step 3: Upgrade Buttons and Inputs
og:image |  https://angulardart-org-dev.firebaseapp.com/codelabs/angular_components/images/cartoon.jpeg?3
og:description | Add material buttons and inputs to your app.

### AngularDart page, e.g., tutorial

Meta Tag | Value
-- | --
og:url | https://angulardart-org-dev.firebaseapp.com/angular/tutorial/toh-pt2
og:type | website
og:title | Master/Detail
og:image |  https://angulardart-org-dev.firebaseapp.com/assets/angulardart.png
og:description | Build a master/detail page with a list of heroes.

### CODE REVIEW

**Comment by chalin** [comment]
Here's what actual shares might look like for the pages listed above:
> screen shot 2017-07-07 at 1 43 47 pm

> screen shot 2017-07-07 at 1 43 39 pm

> screen shot 2017-07-07 at 1 43 32 pm

> screen shot 2017-07-07 at 1 43 21 pm

**Comment by kwalrath** [comment]
@filiph could you also review this PR? I believe this would take the place of #795.

**Comment by filiph** [comment]
Thanks for doing this! The share images now look really good.

I have the same questions as Kathy, though. @chalin, could you elaborate on the changes Kathy points out?

File: _config.yml [hunk 1]
```diff
@@ -28,57 +26,42 @@ api: "https://www.dartdocs.org/documentation"
...
-    permalink: /guides/:path
   codelabs:
     output: true
-    permalink: /codelabs/:path
 
 defaults:
 - scope:
     path: ""
-    type: pages
+    # type: pages
```
Comment (chalin) [inline _config.yml]: This makes the default layout apply to all pages (unless specified otherwise), which is what we want.

File: _config.yml [hunk 2]
```diff
@@ -28,57 +26,42 @@ api: "https://www.dartdocs.org/documentation"
...
+    og_type: article
 - scope:
-    path: ""
-    type: guides
+    path: _codelabs/angular_components
+    type: codelabs
   values:
-    layout: guide
-    toc: true
+    # The `image` query parameter is a version number. Change the number after the file has been updated,
```
Comment (chalin) [inline _config.yml]: If the filename changes then we can reset the version number to 0. Otherwise, it should be incremented.

File: _config.yml [hunk 3]
```diff
@@ -2,9 +2,7 @@
 title: webdev.dartlang.org
 description: "Documentation to help you build high-quality web apps, using Dart tools and libraries. Includes AngularDart docs."
 url: https://webdev.dartlang.org
-baseurl: ""
-port: 4001
-permalink: /:title
```
Comment (chalin) [inline _config.yml]: Having `/:categories/:title` is actually what we want as a permalink for all pages. The previous rule forces us to have a specialized permalink entry for each category.

The baseurl is empty and hence isn't needed. The port has been moved into the `_config_dev.yml` file.

File: _config.yml [hunk 4]
```diff
@@ -28,57 +26,42 @@ api: "https://www.dartdocs.org/documentation"
...
+    # The `image` query parameter is a version number. Change the number after the file has been updated,
+    # otherwise the old cached version will be used by Facebook, etc.
+    image: /codelabs/angular_components/images/cartoon.jpeg?3
 - scope:
-    path: ""
+    path: _codelabs/ng2
     type: codelabs
   values:
-    layout: codelab
-    toc: true
```
Comment (chalin) [inline _config.yml]: There were a bunch of custom layouts that essentially just used the default layout as is. I've removed these layouts.

`toc: true` is set for all pages above.

File: src/_codelabs/darrrt/index.md [hunk 5]
```diff
@@ -1,10 +1,8 @@
 ---
-layout: codelab
 title: "Low-Level HTML Pirate Codelab"
 description: "Take your first step to learning Dart fast."
 permalink: /codelabs/darrrt
-snippet_img: images/piratemap.jpg
-
+image: images/piratemap.jpg?3
```
Comment (chalin) [inline src/_codelabs/darrrt/index.md]: After looking at what other folks were doing, it seems common enough to use the `page.image` property to specify the `og:image`.

Generally, we'll want to set the `og:image` for a subtree of the site. Which is what I've done for all of the Angular docs, as well as for the the ACX and NG codelabs, respectively. This file is the only page of the `darrrt` codelab, hence I've put the `image` property definition here.

File: src/_includes/head.html [hunk 6]
```diff
@@ -2,9 +2,10 @@
   <meta charset="utf-8">
   <meta http-equiv="Content-Language" content="en_US" />
   <meta http-equiv="x-ua-compatible" content="ie=edge">
-  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">{% 
+  assign desc = page.description | default: page.excerpt | default: site.description | strip_html | strip_newlines | truncate: 160
+  %}<meta name="description" content="{{desc}}">
```
Comment (chalin) [inline src/_includes/head.html]: Oups! Fixed

Comment (chalin) [inline src/_includes/head.html]: (I hadn't noticed that from the Chrome dev tools element viewer ... but I can see it clearly when I look at the generated files directly. Good catch!)

File: src/_includes/head.html [hunk 7]
```diff
@@ -17,14 +18,24 @@
...
+  <meta name="twitter:description" content="{{desc}}" />
 
   <!-- Open Graph -->
   <meta property="og:title" content="{{ page.title }}" />
-  <meta property="og:description" content="{% if page.description | strip_html | strip_newlines | truncate: 160 %}{{ page.description }}{% else if page.excerpt %}{{ page.excerpt | strip_html | strip_newlines | truncate: 160 }}{% else %}{{ site.description | strip_html | strip_newlines | truncate: 160 }}{% endif %}" />
-  <meta property="og:type" content="website" />
-  <meta property="og:url" content="{{ site.url }}{{ page.url }}" />
-  <meta property="og:image" content="{% if page.image %}{{ page.image }}{% else %}{{ site.url }}{% asset_path 'dart-webdev-logo.png' %}{% endif %}" />
+  <meta property="og:description" content="{{desc}}" />
+  <meta property="og:url" content="{{ page.url | absolute_url }}" />{%
```
Comment (chalin) [inline src/_includes/head.html]: The value of `page.url` is actually the path to the current page. That is piped into `absolute_url` which essentially adds `site.url` to the front of it.

**Comment by chalin** [comment]
Questions answered. Fix done.

**Comment by chalin** [comment]
I've fixed the AngularDart image size to conform to Facebook best practices. A post from an Angular doc page will look like this:
> screen shot 2017-07-07 at 4 15 09 pm

File: _config.yml (hunk 3 above)
Comment (filiph) [inline _config.yml]: This doesn't break any existing links, does it?

Comment (chalin) [inline _config.yml]: Not as far as I can tell. I've done a reasonable manual sampling of the generated site. It all works.

**Pull request merged**